csharp/src/Ice/TypeIdAttribute.cs

```diff
@@ -0,0 +1,92 @@
+        /// <param name="value">The type ID.</param>>
+        public TypeIdAttribute(string value) => Value = value;
+    }
+
+    /// <summary> This class adds extension methods to System.Type.</summary>
+    public static class TypeExtensions
+    {
+        /// <summary>Retrieve the Ice type ID from a type with the attribute Ice.TypeId</summary>
+        /// <param name="type">The class or interface generated by the Slice compiler.</param>
+        /// <returns>The type ID, or null if type does not carry the Ice.TypeId attribute.</returns>
+        public static string? GetIceTypeId(this Type type)
+        {
+            var attributes = type.GetCustomAttributes(typeof(TypeIdAttribute), false);
+            if (attributes.Length == 1 && attributes[0] is TypeIdAttribute typeId)
+            {
+                return typeId.Value;
+            }
+            else
+            {
+                return null; // unexpected when called by the generated code
+            }
+        }
+
+        /// <summary>Retrieve the Ice type ID from a type and from all its base types.
+        /// When type is an interface, it returns the type ID for that interface, plus all its base interfaces, plus
+        /// "::Ice::Object", and these type IDs are returned in alphabetical order.
+        /// When type is a class, it returns the type ID of that class plus the type ID of the base classes. These
+        /// type IDs are sorted from most derived to least derived.</summary>
+        /// <param name="type">The class or interface generated by the Slice compiler.</param>
+        /// <returns>An array of Ice type IDs.</returns>
+        public static string[] GetAllIceTypeIds(this Type type)
+        {
+            if (type.IsInterface)
+            {
+                var result = new List<string>();
+                if (GetIceTypeId(type) is string firstTypeId)
+                {
+                    result.Add(firstTypeId);
+                    var interfaces = type.GetInterfaces();
+                    foreach (var p in interfaces)
+                    {
+                        if (GetIceTypeId(p) is string typeId)
+                        {
+                            result.Add(typeId);
+                        }
+                    }
+                    result.Add("::Ice::Object");
+                    result.Sort(StringComparer.Ordinal);
+                }
+                return result.ToArray();
+            }
+            else if (type.IsClass)
+            {
+                var result = new List<string>();
+                Type? p = type;
+                do
+                {
+                    var typeId = GetIceTypeId(p!);
+                    if (typeId != null)
+                    {
+                        result.Add(typeId);
+                        p = p.BaseType;
+                    }
+                    else
+                    {
+                        p = null;
+                    }
+                } while (p != null);
```

Comment:
Maybe a for loop can make this a bit simpler
```
for(Type? p = type; p != null; p = p.BaseType)
{
    if (GetIceTypeId(p) is string typeId)
    {
        result.Add(typeId);
    }
}
```

Reply:
Fixed as suggested, except with a `break` in the event a type does not have an Ice type ID.